### PR TITLE
Add scripts runner

### PR DESCRIPTION
**Pull request by maciektr**

### CODE REVIEW

File: scarb/src/ops/subcommands.rs [hunk 1]
```diff
@@ -42,3 +69,42 @@ fn find_external_subcommand(cmd: &str, config: &Config) -> Option<PathBuf> {
...
+        (
+            "scarb".to_string(),
+            Rc::new(ExecutableCommand::new(
+                config.app_exe()?.display().to_string(),
+            )) as Rc<dyn ShellCommand>,
+        ),
+    ]);
+    let list = script_definition.parse(args)?;
+
+    // We use `async_compat::Compat` as `deno_task_shell` is tokio based.
```
Comment (mkaput) [inline scarb/src/ops/subcommands.rs:100]: Let's migrate Scarb to Tokio

**Review by mkaput — CHANGES_REQUESTED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -19,6 +19,7 @@ repository = "https://github.com/software-mansion/scarb"
 [workspace.dependencies]
 anyhow = "1.0.66"
 assert_fs = { version = "1.0.9", features = ["color-auto"] }
+async-compat = "0.2.1"
```
[the remark this answers is not in the document]
Comment (maciektr) [reply Cargo.toml:22]: Yes, I will remove this as soon as we migrate to tokio

File: scarb/src/bin/scarb/args.rs [hunk 3]
```diff
@@ -132,12 +130,22 @@ pub enum Command {
...
+    Run(ScriptsRunnerArgs),
 
     /// External command (`scarb-*` executable).
     #[command(external_subcommand)]
     External(Vec<OsString>),
 }
 
+#[derive(Parser, Clone, Debug)]
+#[clap(trailing_var_arg = true)]
+pub struct ScriptsRunnerArgs {
```
Comment (mkaput) [inline scarb/src/bin/scarb/args.rs:143]: how did this survive `#[deny(missing_docs)]` ??? 😱

File: scarb/src/bin/scarb/commands/run.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+use crate::args::ScriptsRunnerArgs;
+use crate::errors::error_with_exit_code;
+use scarb::core::{Config, Workspace};
+use scarb::ops;
+use scarb::ui::Message;
+
+#[tracing::instrument(skip_all, level = "info")]
+pub fn run(args: ScriptsRunnerArgs, config: &Config) -> Result<()> {
+    let ws = ops::read_workspace(config.manifest_path(), config)?;
+    if args.script.is_some() {
```
Comment (mkaput) [inline scarb/src/bin/scarb/commands/run.rs:16]: if let

File: scarb/src/bin/scarb/commands/run.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    }
+}
+
+fn run_script(
+    script: OsString,
+    args: Vec<OsString>,
+    ws: &Workspace,
+    config: &Config,
+) -> Result<()> {
+    let scripts = &ws.current_package()?.manifest.scripts;
```
Comment (mkaput) [inline scarb/src/bin/scarb/commands/run.rs:29]: please use package selector argument as `scarb add` and `scarb rm` do here

File: scarb/src/bin/scarb/args.rs [hunk 6]
```diff
@@ -132,12 +130,22 @@ pub enum Command {
...
+    pub script: Option<OsString>,
```
Comment (mkaput) [inline scarb/src/bin/scarb/args.rs:144]:
```suggestion
    pub script: Option<SmolStr>,
```

Why OsString here? Wouldn't `SmolStr` make life easier? :D

File: scarb/src/bin/scarb/commands/run.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    config.ui().print(ScriptsList(scripts));
+    Ok(())
+}
+
+#[derive(Serialize, Debug)]
+struct ScriptsList(BTreeMap<String, String>);
+
+impl Message for ScriptsList {
+    fn text(self) -> String {
+        let mut text = String::from("Scripts available via `scarb run`:\n");
```
Comment (mkaput) [inline scarb/src/bin/scarb/commands/run.rs:63]: you can `write!` to `&mut String`

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/bin/scarb/args.rs (hunk 3 above)
Comment (maciektr) [reply scarb/src/bin/scarb/args.rs:143]: 😱

File: scarb/src/bin/scarb/args.rs [hunk 8]
```diff
@@ -132,12 +131,25 @@ pub enum Command {
...
     External(Vec<OsString>),
 }
 
+/// Arguments accepted by the `run` command.
+#[derive(Parser, Clone, Debug)]
+#[clap(trailing_var_arg = true)]
+pub struct ScriptsRunnerArgs {
+    pub script: Option<SmolStr>,
+    #[command(flatten)]
+    pub packages_filter: PackagesFilter,
```
Comment (mkaput) [inline scarb/src/bin/scarb/args.rs:148]: please ensure that `scarb run script -p package` goes to package filter, but `scarb run script -- -p package` goes as an argument to this script

File: scarb/src/bin/scarb/args.rs [hunk 9]
```diff
@@ -132,12 +131,25 @@ pub enum Command {
...
+    #[clap(allow_hyphen_values = true)]
+    pub args: Vec<OsString>,
```
Comment (mkaput) [inline scarb/src/bin/scarb/args.rs:150]: document

**Review by mkaput — CHANGES_REQUESTED** [review]

File: scarb/src/bin/scarb/commands/run.rs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    ws: &Workspace,
+    config: &Config,
+) -> Result<()> {
+    let script_definition = package.manifest.scripts.get(&script).ok_or_else(|| {
+        anyhow!(formatdoc! {r#"
+            missing script `{}`
+
+            To see a list of scripts, run:
+                scarb run
+            "#, script})
```
Comment (mkaput) [inline scarb/src/bin/scarb/commands/run.rs:39]:
```suggestion
        anyhow!(formatdoc! {r#"
            missing script `{script}`

            To see a list of scripts, run:
                scarb run
        "#})
```

**Review by mkaput — APPROVED** [review]
🔥

**Pull request merged**